driver/src/main/java/org/dcache/SapphireDriver.java

```diff
@@ -82,7 +101,198 @@ public void flush(Iterable<FlushRequest> requests)
     @Override
     public void stage(Iterable<StageRequest> requests)
     {
-        throw new UnsupportedOperationException("Not implemented");
+        LOGGER.debug("Stage triggered");
+
+        for(StageRequest stageRequest : requests) {
+            try {
+                stageRequest.activate().get();
+                stageRequestQueue.add(stageRequest);
+                LOGGER.debug("Added {} to stageRequestQueue", stageRequest.getFileAttributes().getPnfsId());
+            } catch (ExecutionException | InterruptedException e) {
+                Throwable t = Throwables.getRootCause(e);
+                LOGGER.error("Failed to activate request {}", t.getMessage());
+                stageRequest.failed(e);
+            }
+        }
+        LOGGER.debug("Length of stageRequestQueue: {}", stageRequestQueue.size());
+    }
+
+    private Checksum calculateAdler32(File file) throws IOException {
+        Adler32 newChecksum = new Adler32();
+
+        byte [] fileArray = FileUtils.readFileToByteArray(file);
+        newChecksum.engineUpdate(fileArray, 0, fileArray.length);
+
+        return new Checksum(ChecksumType.ADLER32, newChecksum.engineDigest());
+    }
+
+    private Checksum calculateMd5(File file) throws NoSuchAlgorithmException, FileNotFoundException { // TODO Needs testing
+        MessageDigest md;
+        md = MessageDigest.getInstance("MD5");
+        DigestInputStream din = new DigestInputStream(new FileInputStream(file), md);
+        LOGGER.debug("Value of DigestInputStream MD5: {}", din);
+
+        if (md == null) {
+            return null;
+        } else {
+            return new Checksum(ChecksumType.MD5_TYPE, md.digest());
```

Comment:
this is wrong, as DigestInputStream calculates digest on `read`. Thus `md.digest` will always return digest of zero bytes.